From 90b9c3c15bc17c773a7fb25d8edcb37e778fb718 Mon Sep 17 00:00:00 2001
From: Karol Herbst <kherbst@redhat.com>
Date: Mon, 3 Aug 2026 21:51:32 +0200
Subject: [PATCH] nak/instr_sched_prepass: immediately schedule instructions if
 we have a carry in use

There seems to have been a bug lurking that could made the prepass
scheduler reorder instructions pairs writing to/reading from the carry
flag. However we don't support spilling it, so for now just force the
prepass scheduler to keep the order of those pairs instact.

In the future we should make the prepass scheduler a bit more smarter or
teach RA how to spill carries, but that seems a bit too much for a bug fix.

Fixes: d470487004f ("nak/instr_sched_prepass: Take predicate spilling into account when scheduling instrucitons")
Fixes: b55b8da012d ("nak: Add a prepass instruction scheduler")
Closes: https://gitlab.freedesktop.org/mesa/mesa/-/work_items/15989
---
 src/nouveau/compiler/nak/opt_instr_sched_prepass.rs | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/src/nouveau/compiler/nak/opt_instr_sched_prepass.rs b/src/nouveau/compiler/nak/opt_instr_sched_prepass.rs
index 854f061b0172..0590e1aeef0d 100644
--- a/src/nouveau/compiler/nak/opt_instr_sched_prepass.rs
+++ b/src/nouveau/compiler/nak/opt_instr_sched_prepass.rs
@@ -488,9 +488,15 @@ impl<'a> GenerateOrder<'a> {
             }
 
             // Pick an instruction to schedule
-            let next_idx = if used_regs[RegFile::GPR]
-                <= thresholds.heuristic_threshold
-                && used_regs[RegFile::Pred] < self.max_regs[RegFile::Pred]
+            // If we have a used carry, schedule everything immediately until
+            // we don't.
+            // TODO: We can't spill carries at the moment, so we can't reorder
+            //       anything between instructions writing to the carry and the
+            //       ones reading from it. In practice those come in pairs, so
+            //       it shouldn't really matter all that much for now.
+            let next_idx = if used_regs[RegFile::Carry] > 0
+                || (used_regs[RegFile::GPR] <= thresholds.heuristic_threshold
+                    && used_regs[RegFile::Pred] < self.max_regs[RegFile::Pred])
             {
                 let ReadyInstr { index, .. } = ready_instrs.pop_last().unwrap();
                 index
-- 
GitLab
